### DISCUSSION TITLE
Drop support for Python 2.7

### DISCUSSION THREAD

**Discussion by olafmersmann** [numbbo/coco#2157] [Polls]
There are lots of places where we have Python 2/3 compatibility crutches in the code. 
I don't think these are necessary anymore and we should remove them and commit ourselves to something like Python 3.6 or maybe even 3.8 as the minimum required version. 
No need to rip out all old code, but anytime we touch a file, we could simplify it by removing the `from __future__` imports etc.

Currently we have
* 15 files in `code-experiments` that import from `__future__`
* 45 files in `code-postprocessing` that import from `__future__`
* 16 files in `code-preprocessing` that import from `__future__`
* and our venerable `do.py` script also import from `__future__`.

**Comment by nikohansen** [comment]
I generally agree. I also have changed the requirement in the readme from Python 2.6 to 3.6.

In particular, the `cocopp` module shouldn't need to support Python 2 anymore.

For `cocoex` and `do.py`, compatibility may be useful under rare circumstances, so I am not sure we need to intentionally drop it, though it may suffice to have a final compatible release.

Unfortunately, I don't even know how to install Python < 3.8 on my current laptop, so I can't even test any older version out-of-the-box.

**Reply by olafmersmann** [reply]
I just ran `do.py test-python` using the last Python 2.7 release on Linux and it didn't throw any obvious errors. Here's the conda environment definiton I used:
```yaml
name: coco-py2
channels:
  - conda-forge
  - defaults
dependencies:
  - python<3.0
  - cython
  - matplotlib
  - numpy
  - pip
  - scipy
  - setuptools
```

Nevertheless, looking at the [Status of Python Versions](https://devguide.python.org/versions/), I would be in favor of silently dropping support for anything older than at least version 3.7 (3.8 is  available on RHEL 7 which is the oldest Linux distribution I would be willing to support). Even if your operating system does not ship with a recent version of Python, you can easily install one using Anaconda.
